cow_instrument/DetectorInfo.h

```diff
@@ -260,27 +367,34 @@ template <typename InstTree> void DetectorInfo<InstTree>::initL2() {
   const std::vector<double> &pathLengths =
       m_pathComponentInfo.const_pathLengths();
 
+  const size_t scanCount = m_durations->size();
+
   // Loop over all detector indexes. We will have a path for each.
   for (size_t detectorIndex = 0; detectorIndex < m_nDetectors;
        ++detectorIndex) {
 
-    auto detectorPos = (*m_positions)[detectorIndex];
-    size_t i = 0;
-    const Path &path = (*m_l2Paths)[detectorIndex];
-    if (path.size() < 1) {
-      throw std::logic_error("Cannot have a L2 specified with less than 1 path "
-                             "components (sample).");
-    }
-    double l2 = pathLengths[path[i]];
-
-    // For each detector-l2-path calculate the total neutronic length
-    for (i = 1; i < path.size(); ++i) {
-      l2 += distance(entryPoints[path[i]], exitPoints[path[i - 1]]);
-      l2 += pathLengths[path[i]];
+    for (size_t timeIndex = 0; timeIndex < scanCount; ++timeIndex) {
+
+      auto detectorPos =
+          (*m_positions)[(*m_linearIndexMap)[detectorIndex][timeIndex]];
+      size_t i = 0;
+      const Path &path = (*m_l2Paths)[detectorIndex];
+      if (path.size() < 1) {
+        throw std::logic_error(
+            "Cannot have a L2 specified with less than 1 path "
+            "components (sample).");
+      }
+      double l2 = pathLengths[path[i]];
+
+      // For each detector-l2-path calculate the total neutronic length
+      for (i = 1; i < path.size(); ++i) {
+        l2 += distance(entryPoints[path[i]], exitPoints[path[i - 1]]);
+        l2 += pathLengths[path[i]];
+      }
+      l2 += distance(exitPoints[path[i - 1]], detectorPos);
+
+      (*m_l2)[(*m_linearIndexMap)[detectorIndex][timeIndex]] = l2;
```

Comment:
Looks good for now. I think it might be possible to improve and simplify this slightly for the final implementation:

- In most cases all detectors have the same `Path` (do I remember correctly? Default would just contain the sample?).
- change `m_l2Paths` to hold only distinct paths.
- have vector of indices that assigns a path to each detector (or each detector position in the scanning case).

Then above loop would just be

```cpp
for (size_t i = 0; i < scanCount * nDetectors; ++i) {
  auto &detectorPos = m_positions[i];
  const Path &path = m_l2Paths[m_pathMap[i]];
  // ...
}
```

This removes the need for nested loops and the index lookup (replacing it with the slightly simpler path lookup).